lib/calculator/lexer.rb

```diff
@@ -176,8 +176,255 @@ def do_right_parenthesis
+      def do_whitespace
+        [:error]
+      end
+
+      # Handles the end of input
+      def do_end_of_input
+        [:error]
+      end
+
+      # Handles a digit
+      def do_digit(d)
+        [:error]
+      end
+
+      # Handles <tt>-</tt>
+      def do_hyphen
+        [:error]
+      end
+
+      # Handles <tt>.</tt>
+      def do_full_stop
+        [:error]
+      end
+
+      # Handles <tt>+</tt>
+      def do_plus
+        [:error]
+      end
+
+      # Handles <tt>*</tt>
+      def do_asterisk
+        [:error]
+      end
+
+      # Handles <tt>/</tt>
+      def do_slash
+        [:error]
+      end
+
+      # Handles <tt>^</tt>
+      def do_caret
+        [:error]
+      end
+
+      # Handles <tt>(</tt>
+      def do_left_parenthesis
+        [:error]
+      end
+
+      # Handles <tt>)</tt>
+      def do_right_parenthesis
+        [:error]
+      end
+    end
+
+
     class DefaultState < State
       include Singleton
+      def do_whitespace
+        [self]
+      end
+
+      def do_end_of_input
+        [:accept]
+      end
+
+      def do_plus
+        [self, AddOpToken.instance]
+      end
+
+      def do_hyphen
+        [DashState.new]
+      end
+
+      def do_asterisk
+        [self, MultiplyOpToken.instance]
+      end
+
+      def do_slash
+        [self, DivideOpToken.instance]
+      end
+
+      def do_caret
+        [self, ExponentOpToken.instance]
+      end
+
+      def do_left_parenthesis
+        [self, LeftParenthesisToken.instance]
+      end
+
+      def do_right_parenthesis
+        [self, RightParenthesisToken.instance]
+      end
+
+      def do_digit(d)
+        [IntegerState.new(d)]
+        end
+
+    class IntegerState < State
+
+      def do_digit(d)
+        @digits += d
+        [self]
+      end
+
+      def do_whitespace
+        [DefaultState.instance, IntegerToken.new(@digits)]
+      end
+
+      def do_end_of_input
+        [:accept, IntegerToken.new(@digits)]
+      end
+
+      def do_full_stop
+        @digits += '.'
+        [DecimalState.new(@digits)]
+      end
+
+      def do_plus
+        [DefaultState.instance, IntegerToken.new(@digits), AddOpToken.instance]
+      end
+
+      def do_hyphen
+        [DefaultState.instance, IntegerToken.new(@digits), SubtractOpToken.instance]
+      end
+
+      def do_slash
+        [DefaultState.instance, IntegerToken.new(@digits), DivideOpToken.instance]
+      end
+
+      def do_asterisk
+        [DefaultState.instance, IntegerToken.new(@digits), MultiplyOpToken.instance]
+      end
+
+      def do_caret
+        [DefaultState.instance, IntegerToken.new(@digits), ExponentOpToken.instance]
+      end
+
+      def do_left_parenthesis
+        [DefaultState.instance, IntegerToken.new(@digits), LeftParenthesisToken.instance]
+      end
+
+      def do_right_parenthesis
+        [DefaultState.instance, IntegerToken.new(@digits), RightParenthesisToken.instance]
+      end
+
+    end
+
+    class DashState < State
+
```

Comment:
You never initialise `@dash` so you will get an exception about not being able to call `+` on `nil`